picamera2/encoders/v4l2_encoder.py

```diff
@@ -214,32 +217,35 @@ def encode(self, stream, request):
         """
         # Don't start encoding if we don't have an output handle
         # as the header seems only to be sent with the first frame
-        if self._output is None:
-            return
-        if self.vd is None or self.vd.closed:
-            return
-        cfg = stream.configuration
-        fb = request.request.buffers[stream]
-        fd = fb.planes[0].fd
-        request.acquire()
-
-        buf = v4l2_buffer()
-        # fb.metadata.timestamp is in nanoseconds, so convert to usecs
-        timestamp_us = self._timestamp(fb)
-
-        # Pass frame to video 4 linux, to encode
-        planes = v4l2_plane * VIDEO_MAX_PLANES
-        planes = planes()
-        buf.type = V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE
-        buf.index = self.buf_available.get()
-        buf.field = V4L2_FIELD_NONE
-        buf.memory = V4L2_MEMORY_DMABUF
-        buf.length = 1
-        buf.timestamp.secs = timestamp_us // 1000000
-        buf.timestamp.usecs = timestamp_us % 1000000
-        buf.m.planes = planes
-        buf.m.planes[0].m.fd = fd
-        buf.m.planes[0].bytesused = cfg.frame_size
-        buf.m.planes[0].length = cfg.frame_size
-        fcntl.ioctl(self.vd, VIDIOC_QBUF, buf)
-        self.buf_frame.put(request)
+        with self.lock:
```

Comment:
Just wondering, like you say if you moved the lock for 'stop' into the Encoder class.  Would we then want this self.lock initiated in the picamera class possibly before it calls 'encode'?

Reply:
Around this call - https://github.com/raspberrypi/picamera2/blob/next/picamera2/picamera2.py#L1159

Reply:
I expect the lock would go in the Encoder base class. Then everyone always calls a base class method (which deals with the lock), and then calls the more internal "do the work" derived method. Here we're saying that "encode" and "stop" need this behaviour, actually I wonder if "start" does too (see my next comment).